core/controllers/base_test.py

```diff
@@ -95,16 +114,72 @@ def test_requests_for_invalid_paths(self):
         response = self.testapp.get('/')
         self.assertEqual(response.status_int, 302)
         self.assertIn('splash', response.headers['location'])
 
-        # Login and assert that there is a redirect
-        self.login('user@example.com')
+    def test_root_redirect_rules_for_logged_in_learners(self):
+        self.login(self.TEST_LEARNER_EMAIL)
+
+        # Since no exploration has been created, going to '/' should redirect
+        # to the library page.
+        response = self.testapp.get('/')
+        self.assertEqual(response.status_int, 302)
+        self.assertIn('library', response.headers['location'])
+        self.logout()
+
+    def test_root_redirect_rules_for_logged_in_creators(self):
+        self.login(self.TEST_CREATOR_EMAIL)
+        creator_user_id = self.get_user_id_from_email(self.TEST_CREATOR_EMAIL)
+        exploration_id = '0_en_test_exploration'
+        self.save_new_valid_exploration(
+            exploration_id, creator_user_id, title='Test',
+            category='Test', language_code='en')
+
+        # Since atleast one exploration has been created, going to '/' should
+        # redirect to the dashboard page.
+        response = self.testapp.get('/')
+        self.assertIn('dashboard', response.headers['location'])
+        exp_services.delete_exploration(creator_user_id, exploration_id)
+        self.logout()
+        self.login(self.TEST_CREATOR_EMAIL)
+        response = self.testapp.get('/')
+
+        # Even though the exploration is deleted, the user is still redirected
+        # to the dashboard. This is because deleted explorations are still
+        # associated with their creators.
+        self.assertEqual(response.status_int, 302)
+        self.assertIn('dashboard', response.headers['location'])
+
+    def test_root_redirect_rules_for_logged_in_editors(self):
+        self.login(self.TEST_CREATOR_EMAIL)
+        creator_user_id = self.get_user_id_from_email(self.TEST_CREATOR_EMAIL)
+        editor_user_id = self.get_user_id_from_email(self.TEST_EDITOR_EMAIL)
+        exploration_id = '1_en_test_exploration'
+        self.save_new_valid_exploration(
+            exploration_id, creator_user_id, title='Test',
+            category='Test', language_code='en')
+        rights_manager.assign_role_for_exploration(
+            creator_user_id, exploration_id, editor_user_id,
+            rights_manager.ROLE_EDITOR)
+        self.logout()
+        self.login(self.TEST_EDITOR_EMAIL)
+        exp_services.update_exploration(
+            editor_user_id, exploration_id, [{
+                'cmd': 'edit_exploration_property',
+                'property_name': 'title',
+                'new_value': 'edited title'
+            }, {
+                'cmd': 'edit_exploration_property',
+                'property_name': 'category',
+                'new_value': 'edited category'
+            }], 'Change title and category')
+
+        # Since user has edited one exploration created by another user
```

Comment:
I apologize for my English. It is not my first language. I didn't double check my comments in haste. I will try to do better in future.